src/search/parser/lexical_analyzer.cc

```diff
@@ -29,6 +29,8 @@ static vector<pair<TokenType, regex>> construct_token_type_expressions() {
         {TokenType::INTEGER,
          R"([+-]?(infinity|\d+([kmg]\b)?))"},
         {TokenType::BOOLEAN, R"(true|false)"},
+        // TODO: support quoted strings.
+        {TokenType::STRING, R"("([^"]*)\")"},
```

Comment:
I don't understand why the first " in the regexp appears to be unquoted, but the second one is preceded by a backslash. Shouldn't they be symmetric?

In my view, we need to resolve the TODO before merging this. I wouldn't overdo it, but I think we should support three types of escapes: `\\` for a backslash, `\"` for a double quote, and `\n` for a newline.I think the regexp could be something along the lines of:

```"(\\|\"|\n|^["\])*"```

except that I assume some characters in there will need escaping and we might want to exclude more items in the final option. But I hope the structure is clear.

Reply:
I started working on this a while back and wanted to do proper escaping/unescaping with a function to parse an escaped string. I ran into another problem during this, where our error messages want to print the relevant part of the command line. The same holds for log messages for heuristic values (i.e., in `astar(lmcut())` we want to print `lmcut()` when reporting its heuristic value). If there is a string option, and we parse it, we should escape it again when we print it, similar to `str()` and `repr()` in Python, where we currently only have `str` and not `repr`.

I have a partial implementation of this but didn't get around to finishing it yet, and we didn't pull this story in the current sprint, so I'm not focusing on it.

Reply:
I added `escape` and `unescape` functions to `utils/strings` and use them to create the escaped version of string tokens where we dump values for debugging. I first implemented the parsing of quoted strings manually but the regex suggested by @maltehelmert is a better solution. There is still some specialized handling of string tokens, though: for example, we convert all other tokens to lower case and don't do so for strings. Also, we have to unescape string tokens and remove the surrounding quotation marks.

Reply:
It's normal for every token type to involve some processing to understand the "value", and I wouldn't consider the unescaping to be "specialized handling" in this regard. For example, numbers needed to be converted from strings to numbers, and "false"/"true" need to be converted to bools. This would be the normal place for the evaluation of the string literal to live. Do they not live in the same/a comparable place in the code?

Regarding conversion to lower-case: I think it makes a bit more sense to have the tokens that care about folding to lower-case support both cases (and for example the symbol token type to do a conversion) rather than consider the string literal to be an exception to a general rule, but if you prefer implementing the string literals as an exception because it affects fewer places, that's fine. We might want to eventually make things case-sensitive though, to further reduce the differences to the Python code.

Reply:
I previously had in mind that in an input string `"\"tasty\" cheese"`, the token would be a single string token with the value `"tasty" cheese`, but after what you commented, I now changed this view to see this as a string token with content `"\"tasty\" cheese"`. This stays the value of the token until the stage where we construct C++ objects from our AST. In that stage we would then unescape the token `"\"tasty\" cheese"` to the `std::string` `"tasty" cheese`. This is the same stage where we also interpret integer tokens like `infinity` or `10k`, so I think this fits nicely. The whole diff becomes a bit nicer this way, so I think this is the way to go.

I also moved the code that transforms the token to lower case from the parsing code to the constructor of `Token`. The token knows its type and decides based on that if it should transform its value to lower case. I think this is also a nice way of handling things and makes the parsing code uniform for all token types.

This is now ready to merge from my point of view.